### PR TITLE
Voting Contract: Add list_votes_by_voter query

### PR DESCRIPTION
**Pull request by ueco-jb**
closes https://github.com/confio/poe-contracts/issues/78

### CODE REVIEW

File: packages/voting-contract/src/state.rs [hunk 1]
```diff
@@ -267,6 +267,7 @@ pub const PROPOSAL_COUNT: Item<u64> = Item::new("proposal_count");
 
 // multiple-item map
 pub const BALLOTS: Map<(u64, &Addr), Ballot> = Map::new("votes");
+pub const BALLOTS_BY_VOTER: Map<(&Addr, u64), Ballot> = Map::new("votes_by_voter");
```
Comment (uint) [inline packages/voting-contract/src/state.rs:270]: Is there a storage option that would let us store all the ballots in a single "map" with two indexes? Like the `IndexedMap` from `storage-plus`?

**Review by uint — APPROVED** [review]
Looks good overall!

File: packages/voting-contract/src/state.rs (hunk 1 above)
Comment (ueco-jb) [reply packages/voting-contract/src/state.rs:270]: Good idea, I think it allows indexing by multiple keys.
I'll check that out.

Comment (ethanfrey) [reply packages/voting-contract/src/state.rs:270]: This is how we [do it in trusted-circle](https://github.com/confio/tgrade-contracts/blob/main/contracts/tgrade-trusted-circle/src/state.rs#L613-L614) as well. I think this is fine for this PR as they are only written once and never updated.

Using IndexedMap for this one is a nice improvement, but can be pulled out to another issue (either for 0.6 or 0.7), and trusted circle should also be updated then. Can you make such an issue and link it? Add it to 0.6 "patch net stectch" milestone for starters

Comment (ueco-jb) [reply packages/voting-contract/src/state.rs:270]: Done. https://github.com/confio/poe-contracts/issues/83

File: packages/voting-contract/src/lib.rs [hunk 2]
```diff
@@ -136,7 +137,13 @@ where
             vote,
         }),
     })?;
-
+    BALLOTS_BY_VOTER.update(deps.storage, (&info.sender, proposal_id), |bal| match bal {
```
Comment (ethanfrey) [inline packages/voting-contract/src/lib.rs:140]: Nit: we don't need to Update here (extra read), as the uniqueness of the key is enforced above.

Also, we could combine these two into [one function "save_ballot"](https://github.com/confio/tgrade-contracts/blob/main/contracts/tgrade-trusted-circle/src/state.rs#L621-L629). Maybe with the "AlreadyVoted" check inside the helper. [Something like this](https://github.com/confio/tgrade-contracts/blob/main/contracts/tgrade-trusted-circle/src/contract.rs#L553-L564)

File: packages/voting-contract/src/state.rs (hunk 1 above)
Comment (ueco-jb) [reply packages/voting-contract/src/state.rs:270]: @ethanfrey 
> I think this is fine for this PR as they are only written once and never updated.

...although they are updated.
https://github.com/confio/poe-contracts/pull/79/files#diff-4658a3ea273c6042143567e7913e908f5166041b5a186a97516f90f6223b9f78R140
Does that change your position here?

File: packages/voting-contract/src/lib.rs [hunk 3]
```diff
@@ -136,7 +137,13 @@ where
...
+        Some(_) => Err(ContractError::AlreadyVoted {}),
+        None => Ok(Ballot {
+            weight: vote_power,
+            vote,
+        }),
+    })?;
```
Comment (ueco-jb) [inline packages/voting-contract/src/lib.rs:146]: @ethanfrey 
> I think this is fine for this PR as they are only written once and never updated.

...although they are updated.

Does that change your position here?

**Review by ethanfrey — APPROVED** [review]
Looks good. A few comments.

I think adding one helper that updates both BALLOTS and BALLOTS_BY_VOTER is simpler than IndexedMap and just as effective. (IndexedMap has some learning curve - best for things that get updated multiple places)

File: packages/voting-contract/src/lib.rs (hunk 2 above)
Comment (ueco-jb) [reply packages/voting-contract/src/lib.rs]: Done, create separate helper and replaced `update` with `save` to reduce read.

File: packages/voting-contract/src/lib.rs [hunk 4]
```diff
@@ -105,6 +105,20 @@ where
         .set_data(cosmwasm_std::to_binary(&resp)?))
 }
 
+fn update_ballot(
```
Comment (ethanfrey) [inline packages/voting-contract/src/lib.rs:108]: Maybe we use this on lines 95-96 as well?
(optional)

**Review by ethanfrey — APPROVED** [review]
lgtm

File: packages/voting-contract/src/lib.rs (hunk 4 above)
Comment (ueco-jb) [reply packages/voting-contract/src/lib.rs:108]: There are two saves vs update and save in heler - so we are back again to one extra read.
Is that okay?

Comment (ueco-jb) [reply packages/voting-contract/src/lib.rs]: Okay, proper helper worked fine.
Done.

**Pull request merged**